libnavigation-core/src/test/java/com/mapbox/navigation/core/reroute/MapboxRerouteControllerTest.kt

```diff
@@ -0,0 +1,333 @@
+    private lateinit var rerouteController: MapboxRerouteController
+
+    @MockK
+    private lateinit var directionsSession: DirectionsSession
+
+    @MockK
+    private lateinit var tripSession: TripSession
+
+    @MockK
+    private lateinit var routeOptionsProvider: RouteOptionsProvider
+
+    @MockK
+    private lateinit var logger: Logger
+
+    @MockK
+    private lateinit var routeOptions: RouteOptions
+
+    @MockK
+    private lateinit var routeCallback: RerouteController.RoutesCallback
+
+    @MockK
+    lateinit var primaryRerouteObserver: RerouteController.RerouteStateObserver
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this, relaxUnitFun = true, relaxed = true)
+        rerouteController = spyk(
+            MapboxRerouteController(
+                directionsSession,
+                tripSession,
+                routeOptionsProvider,
+                logger
+            )
+        )
+    }
+
+    @After
+    fun cleanUp() {
+        assertEquals(RerouteState.Idle, rerouteController.state)
+        // routeCallback mustn't called in current implementation. DirectionSession update routes internally
+        verify(exactly = 0) { routeCallback.onNewRoutes(any()) }
+    }
+
+    @Test
+    fun initial_state() {
+        assertEquals(rerouteController.state, RerouteState.Idle)
+        verify(exactly = 0) { rerouteController.state = any() }
+        verify(exactly = 0) { rerouteController.reroute(any()) }
+        verify(exactly = 0) { rerouteController.interrupt() }
+    }
+
+    @Test
+    fun initial_state_with_added_state_observer() {
+        val added = addRerouteStateObserver()
+
+        assertTrue("RerouteStateObserver is not added", added)
+        verify(exactly = 1) { primaryRerouteObserver.onNewState(RerouteState.Idle) }
+    }
+
+    @Test
+    fun reroute_success() {
+        mockNewRouteOptions()
+        addRerouteStateObserver()
+        val routeRequestCallback = slot<RoutesRequestCallback>()
+        every {
+            directionsSession.requestRoutes(
+                routeOptions,
+                capture(routeRequestCallback)
+            )
+        } returns mockk()
+
+        rerouteController.reroute(routeCallback)
+        routeRequestCallback.captured.onRoutesReady(mockk())
+
+        assertTrue(routeRequestCallback.isCaptured)
+        verify(exactly = 1) {
+            primaryRerouteObserver.onNewState(RerouteState.FetchingRoute)
+        }
+        verify(exactly = 1) {
+            primaryRerouteObserver.onNewState(RerouteState.RouteHasBeenFetched)
+        }
+        verify(exactly = 2) {
+            primaryRerouteObserver.onNewState(RerouteState.Idle)
+        }
+        verifyOrder {
+            primaryRerouteObserver.onNewState(RerouteState.Idle)
+            primaryRerouteObserver.onNewState(RerouteState.FetchingRoute)
+            primaryRerouteObserver.onNewState(RerouteState.RouteHasBeenFetched)
+            primaryRerouteObserver.onNewState(RerouteState.Idle)
+        }
```

Comment:
nit: verifying the order also verifies the count so can choose one

```suggestion
        verifyOrder {
            primaryRerouteObserver.onNewState(RerouteState.Idle)
            primaryRerouteObserver.onNewState(RerouteState.FetchingRoute)
            primaryRerouteObserver.onNewState(RerouteState.RouteHasBeenFetched)
            primaryRerouteObserver.onNewState(RerouteState.Idle)
        }
```

Reply:
from documentation: `Order is important, but not all calls are checked`

anyway, I'm gonna remove checks for `state` property because setter is private from now

Reply:
> from documentation: Order is important, but not all calls are checked

haha why can't i find this? but did find a thread where others discussed https://github.com/mockk/mockk/issues/225 so can see the point. 

still think we can choose one for simpler tests, check order or check count. but up to you, it's a nitpick comment

Reply:
@kmadsen https://github.com/mockk/mockk/blob/master/dsl/common/src/main/kotlin/io/mockk/API.kt#L633 😄 

I think it's ok to keep tests like that, because here is important both options: ordering and count of calls 👍